File: frontend/pda/src/api/cycleCount.ts
// Cycle Count API — PDA operator workflow for physical inventory counting.
// Operators start counts, scan location barcodes, enter counted quantities per SKU/batch,
// and submit the variance report for review.

import client from './client'

// ── Response Types ────────────────────────────────────────────────────────

export interface CycleCountLine {
  id: string
  sku_id: string
  location_id: string
  batch_no: string
  system_qty: number
  counted_qty?: number
  variance?: number
  status: CycleCountLineStatus
  counted_at?: string
  created_at: string
}

export type CycleCountLineStatus = 'pending' | 'counted' | 'reviewed'

export type CycleCountStatus = 'draft' | 'in_progress' | 'pending_review' | 'approved' | 'adjusted' | 'cancelled'

export interface CycleCount {
  id: string
  count_no: string
  warehouse_id: string
  location_id?: string
  zone_id?: string
  status: CycleCountStatus
  counted_by: string
  notes: string
  total_lines: number
  matched_lines: number
  created_at: string
  started_at?: string
  completed_at?: string
  approved_at?: string
  approved_by?: string
  lines?: CycleCountLine[]
}

export interface StartCycleCountRequest {
  warehouse_id: string
  location_id?: string
  zone_id?: string
  counted_by?: string
  notes?: string
}

export interface SubmitLineRequest {
  line_id: string
  counted_qty: number
}

export interface FinalizeCountRequest {
  notes?: string
}

export interface ApproveCountRequest {
  approved_by?: string
  action: 'approve' | 'adjust'
}

// ── API Functions ─────────────────────────────────────────────────────────

export async function startCycleCount(input: StartCycleCountRequest): Promise<CycleCount> {
  const { data } = await client.post<CycleCount>('/cycle-counts', input)
  return data
}

export async function getCycleCount(id: string): Promise<CycleCount> {
  const { data } = await client.get<CycleCount>(`/cycle-counts/${id}`)
  return data
}

export async function listCycleCounts(params?: {
  warehouse_id?: string
  status?: string
}): Promise<CycleCount[]> {
  const { data } = await client.get<{ data: CycleCount[] }>('/cycle-counts', { params })
  return data.data
}

export async function submitLine(cycleCountId: string, input: SubmitLineRequest): Promise<CycleCountLine> {
  const { data } = await client.post<CycleCountLine>(`/cycle-counts/${cycleCountId}/lines`, input)
  return data
}

export async function finalizeCount(id: string, input?: FinalizeCountRequest): Promise<CycleCount> {
  const { data } = await client.post<CycleCount>(`/cycle-counts/${id}/finalize`, input || {})
  return data
}

export async function approveCount(id: string, input: ApproveCountRequest): Promise<CycleCount> {
  const { data } = await client.put<CycleCount>(`/cycle-counts/${id}/approve`, input)
  return data
}

export async function cancelCycleCount(id: string): Promise<CycleCount> {
  const { data } = await client.put<CycleCount>(`/cycle-counts/${id}/cancel`)
  return data
}

File: backend/internal/domain/cycle_count.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// CycleCount represents a physical inventory count for a specific warehouse area.
// It tracks the lifecycle from creation → in_progress → pending_review → approved/adjusted.
type CycleCount struct {
	ID           uuid.UUID        `json:"id"`
	CountNo      string           `json:"count_no"`      // e.g. "CC-20260724-001"
	WarehouseID  uuid.UUID        `json:"warehouse_id"`
	LocationID   *uuid.UUID       `json:"location_id,omitempty"` // Specific location or nil for zone
	ZoneID       *uuid.UUID       `json:"zone_id,omitempty"`     // Specific zone or nil for location
	Status       CycleCountStatus `json:"status"`
	CountedBy    string           `json:"counted_by,omitempty"` // Operator ID
	Notes        string           `json:"notes,omitempty"`
	TotalLines   int              `json:"total_lines"`   // Total count lines
	MatchedLines int              `json:"matched_lines"` // Lines with no variance
	CreatedAt    time.Time        `json:"created_at"`
	StartedAt    *time.Time       `json:"started_at,omitempty"`
	CompletedAt  *time.Time       `json:"completed_at,omitempty"`
	ApprovedAt   *time.Time       `json:"approved_at,omitempty"`
	ApprovedBy   string           `json:"approved_by,omitempty"`
}

// CycleCountLine represents a single counted item (SKU/batch at a location) within a cycle count.
type CycleCountLine struct {
	ID             uuid.UUID           `json:"id"`
	CycleCountID   uuid.UUID           `json:"cycle_count_id"`
	SKUID          uuid.UUID           `json:"sku_id"`
	LocationID     uuid.UUID           `json:"location_id"`
	BatchNo        string              `json:"batch_no,omitempty"`
	SystemQty      float64             `json:"system_qty"`   // What the system says is there
	CountedQty     *float64            `json:"counted_qty,omitempty"` // What the operator counted
	Variance       *float64            `json:"variance,omitempty"`    // counted_qty - system_qty
	Status         CycleCountLineStatus `json:"status"`
	CountedAt      *time.Time          `json:"counted_at,omitempty"`
	CreatedAt      time.Time           `json:"created_at"`
}

// CycleCountStatus represents the lifecycle state of a cycle count.
type CycleCountStatus string

const (
	CycleCountStatusDraft        CycleCountStatus = "draft"         // Created but not started
	CycleCountStatusInProgress   CycleCountStatus = "in_progress"   // Operator is counting
	CycleCountStatusPendingReview CycleCountStatus = "pending_review" // Awaiting supervisor review
	CycleCountStatusApproved     CycleCountStatus = "approved"      // Variance approved, inventory updated
	CycleCountStatusAdjusted     CycleCountStatus = "adjusted"      // Adjusted (override) applied
	CycleCountStatusCancelled    CycleCountStatus = "cancelled"     // Cancelled
)

// CycleCountLineStatus represents the state of an individual count line.
type CycleCountLineStatus string

const (
	CycleCountLineStatusPending  CycleCountLineStatus = "pending"  // Not yet counted
	CycleCountLineStatusCounted  CycleCountLineStatus = "counted"  // Operator has entered count
	CycleCountLineStatusReviewed CycleCountLineStatus = "reviewed" // Reviewed by supervisor
)

// ── CycleCount State Machine ─────────────────────────────────────────────────

// IsTerminal returns true if the cycle count is in a terminal state.
func (cc *CycleCount) IsTerminal() bool {
	return cc.Status == CycleCountStatusApproved ||
		cc.Status == CycleCountStatusAdjusted ||
		cc.Status == CycleCountStatusCancelled
}

// CanTransitionTo checks whether the cycle count can transition from its current
// status to the target status.
//
// Valid transitions:
//
//	draft          → in_progress, cancelled
//	in_progress    → pending_review, cancelled
//	pending_review → approved, adjusted, cancelled
//	approved       → (terminal)
//	adjusted       → (terminal)
//	cancelled      → (terminal)
func (cc *CycleCount) CanTransitionTo(target CycleCountStatus) bool {
	if cc.Status == target {
		return false
	}
	if cc.IsTerminal() {
		return false
	}
	// Any non-terminal status can be cancelled.
	if target == CycleCountStatusCancelled {
		return true
	}

	switch cc.Status {
	case CycleCountStatusDraft:
		return target == CycleCountStatusInProgress
	case CycleCountStatusInProgress:
		return target == CycleCountStatusPendingReview
	case CycleCountStatusPendingReview:
		return target == CycleCountStatusApproved || target == CycleCountStatusAdjusted
	default:
		return false
	}
}

// CanBeStarted returns true if the cycle count can be started.
func (cc *CycleCount) CanBeStarted() bool {
	return cc.Status == CycleCountStatusDraft
}

// CanBeSubmitted returns true if the cycle count can be submitted for review.
func (cc *CycleCount) CanBeSubmitted() bool {
	return cc.Status == CycleCountStatusInProgress
}

// CanBeReviewed returns true if the cycle count is ready for approval.
func (cc *CycleCount) CanBeReviewed() bool {
	return cc.Status == CycleCountStatusPendingReview
}

// HasVariances returns true if any line has a non-zero variance.
func (cc *CycleCount) HasVariances(lines []*CycleCountLine) bool {
	for _, l := range lines {
		if l.Variance != nil && *l.Variance != 0 {
			return true
		}
	}
	return false
}

// ── CycleCountLine Methods ───────────────────────────────────────────────────

// IsCounted returns true if the line has been counted.
func (l *CycleCountLine) IsCounted() bool {
	return l.Status == CycleCountLineStatusCounted || l.Status == CycleCountLineStatusReviewed
}

// SetCountedQty sets the counted quantity and computes variance.
func (l *CycleCountLine) SetCountedQty(qty float64) {
	l.CountedQty = &qty
	v := qty - l.SystemQty
	l.Variance = &v
	l.Status = CycleCountLineStatusCounted
	now := time.Now()
	l.CountedAt = &now
}
